sway-core/src/convert_parse_tree.rs

```diff
@@ -505,22 +512,26 @@ fn item_enum_to_enum_declaration(
     item_enum: ItemEnum,
 ) -> Result<EnumDeclaration, ErrorEmitted> {
     let span = item_enum.span();
+    let variants = item_enum
+        .fields
+        .into_inner()
+        .into_iter()
+        .enumerate()
+        .map(|(tag, type_field)| type_field_to_enum_variant(ec, type_field, tag))
+        .collect::<Result<Vec<_>, _>>()?;
+    if variants.iter().any(|variant| {
+       matches!(&variant.r#type, TypeInfo::Custom { name, ..} if name == &item_enum.name)
```

Comment:
Yeah, nah.  It's just looking for direct child to parent relationships.  Fair call.

Reply:
Aww, man, I have just made a fix for mutual recursion, even between structs and enums and back.  But the generic stuff is tough, since it'd have to be done after type checking.

Reply:
Can we maybe postpone the generics recursive check until we do Emily's refactor?  According to her plan the dependency ordering will be completely different then, and so the recursive check will be too... although, I imagine we _will_ support recursive calls and types then...